### PR TITLE
Expose metrics to non-admins

### LINKED ISSUE #3252: Option to expose command line and job metrics
1. Log into a Galaxy as a non-administrator, e.g. https://usegalaxy.org
2. Click on the "i" (view details) icon of a tool output
3. You get lots of information in first block of entries, from Name to UUID.
4. Log into your Galaxy as an administrator
5. Click on the "i" (view details) icon of a tool output
6. You get lots of extra information after the UUID:

 - Full Path
 - Job Command-Line
 - Job Runtime (Wall Clock)
 - Cores Allocated
 - Job Start Time
 - Job End Time

There is an option in ``config/galaxy.ini`` to allow non-admins to see the full path:

```
# This option allows users to see the full path of datasets via the "View
# Details" option in the history.  Administrators can always see this.
#expose_dataset_path = False
```

This is implemented in ``templates/show_params.mako`` here:

https://github.com/galaxyproject/galaxy/blob/71cea6604d43c5fe6215f5656462ba6c1af69bb6/templates/show_params.mako#L169

```
        %if trans.user_is_admin() or trans.app.config.expose_dataset_path:
            <tr><td>Full Path:</td><td>${hda.file_name | h}</td></tr>
        %endif
        %if job and job.command_line and trans.user_is_admin():
            <tr><td>Job Command-Line:</td><td>${ job.command_line | h }</td></tr>
        %endif
        %if job and trans.user_is_admin():
            <% job_metrics = trans.app.job_metrics %>
            %for metric in job.metrics:
                <% metric_title, metric_value = job_metrics.format( metric.plugin, metric.metric_name, metric.metric_value ) %>
                <tr><td>${ metric_title | h }</td><td>${ metric_value | h }</td></tr>
            %endfor
        %endif
```

Original commit https://github.com/galaxyproject/galaxy/commit/68c9d80fe8b5d3aeb5dc47d78a377a0c5dab1863 by @dannon

I would like two extra options, something like ``expose_job_command_line`` and ``expose_job_metrics``, to make these fields available to non-admins.

There is a security concern as the command line will reveal the dataset paths, but if you are using ``expose_dataset_path=True`` then showing the command line too would help some of your power users (e.g. debugging why a job failed, or how it differs from a similar job run by hand at the command line).

I see little downside to exposing the job metrics, and this has the upside of helping users review jobs later (e.g. if left to run overnight they can't really know how long/slow/expensive each step was).

See also https://github.com/galaxyproject/galaxy/commit/95baa6950ebd6bf3ba917f90d41978ebc294e431 by @jmchilton which added the metric report.

(I'm happy to attempt a pull request if this is likely to be accepted)

### PR DESCRIPTION
**Pull request by hexylena**
Fixes #3252 and #2954. Has commits from #3340 so that should be merged first.

### CODE REVIEW

**Comment by hexylena** [comment]
Rebased to pull in lint fix.

File: config/galaxy.ini.sample [hunk 1]
```diff
@@ -1113,6 +1113,13 @@ use_interactive = True
 # Details" option in the history.  Administrators can always see this.
 #expose_dataset_path = False
 
+# This option allows users to see the job metrics (except for environment
+# variables).
+#expose_job_metrics = False
+
+# This option allows users to see the built command line.
+#expose_job_command_line = False
```
Comment (jmchilton) [inline config/galaxy.ini.sample]: I feel like we should reuse ``expose_dataset_path`` for ``expose_job_command_line`` and we could rename ``expose_job_metrics`` to ``expose_potentially_sensitive_job_metrics`` so we could one day default to exposing the ``core`` metrics for instance - I feel like this one should just be displayed to everyone regardless of setting up extra things.

Comment (hexylena) [reply config/galaxy.ini.sample]: @peterjc you requested the granularity. Thoughts?

**Comment by peterjc** [comment]
This is pretty much exactly what I was picturing implementing, so +1.

Regarding John's suggestion, showing the command line implicitly gives away the (input and output) dataset paths, so there is some logic in using a single setting for both.

**Comment by hexylena** [comment]
Okie dokie. Simplified.

**Comment by bgruening** [comment]
@erasche @peterjc is this just the path which we care about? Why do we not replace the path with a placeholder for non-admins? Every job should know the job_working_directory so we can simply .replace('job_working_directory', '[hidden path]'). If this is still to insecure because of databases and data from loc files we could replace all path with a regex?

**Comment by hexylena** [comment]
Devil's advocate, why do we hide paths at all? What real, *actionable intelligence* could a malicious cluster user (i.e. worst case scenario) obtain from knowing that some files exist in some paths?

**Comment by jmchilton** [comment]
@erasche The default deployment should reveal as little information as necessary about the configuration of the deployment. This is a fairly traditional approach to securing web applications. Certainly the attacks I can imagine in your described situation are contrived - but if we succeed in pushing Galaxy into more and more sensitive areas of the field there is going to be continued pushes to lock down information exposure. We've already received some in the past.

@bgruening This would be incomplete unless the regex was so expansive that the result would be quite confusing. I prefer the current approach.

I'm +1 now that these changes have been made - can everyone agree this is at least a step forward?

**Comment by hexylena** [comment]
@jmchilton fair enough. (xref [OWASP](https://www.owasp.org/index.php/Full_Path_Disclosure)). I just like to push back on security requirements. This one has always felt like a bit of security through obscurity, but I can certainly recognise there are organisations who are obsessed with passing their automated vulnerability reports.

**Comment by peterjc** [comment]
Thanks all - this is a useful step forward and does not expose anything more by default.

**Comment by jmchilton** [comment]
Thanks for the input all! Thanks for the implementation @erasche - very nice work!

**Pull request merged**